### PR TITLE
feat: Add `Guild.delete_auto_moderation_rule`

### PR DESCRIPTION
**Pull request by llamaair**
## Summary

Add a function to delete auto moderation rules either by name or id.

## Information

- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting,
      examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [ ] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.
- [x] I have updated the changelog to include these changes.

### CODE REVIEW

**Review by NeloBlivion — COMMENTED** [review]
Nowhere else in the library do we include a `name` kwarg for deletions (roles, emojis, guilds, channels etc.)... it shouldn't be included here either.
(We also already have [`AutoModRule.delete()`](https://docs.pycord.dev/en/master/api/models.html#discord.AutoModRule.delete), but having a method on guild is understandable to avoid an extra fetch.)

**Comment by llamaair** [comment]
> Nowhere else in the library do we include a `name` kwarg for deletions (roles, emojis, guilds, channels etc.)... it shouldn't be included here either.

Ok, I'l remove th name kwarg 👍

**Review by JustaSqu1d — APPROVED** [review]

File: discord/guild.py [hunk 1]
```diff
@@ -3816,3 +3816,58 @@ async def create_auto_moderation_rule(
             self.id, payload, reason=reason
         )
         return AutoModRule(state=self._state, data=data)
+
+    async def delete_auto_moderation_rule(
+        self,
+        *,
+        id: int | None = None,
+        name: str | None = None,
```
Comment (Lulalaby) [inline discord/guild.py:3824]:
```suggestion
```

File: discord/guild.py [hunk 2]
```diff
@@ -3816,3 +3816,58 @@ async def create_auto_moderation_rule(
...
+    ) -> AutoModRule:
+        """
+        Deletes an auto moderation rule.
+
+        Parameters
+        ----------
+        id: Optional[int]
+            The ID of the auto moderation rule.
+        name: Optional[str]
+            The name of the auto moderation rule.
```
Comment (Lulalaby) [inline discord/guild.py:3835]:
```suggestion
```

File: discord/guild.py [hunk 3]
```diff
@@ -3816,3 +3816,58 @@ async def create_auto_moderation_rule(
...
+            The reason for deleting the rule. Shows up in the audit log.
+
+        Raises
+        ------
+        ValueError
+            If neither 'id' nor 'name' is provided.
+        ValueError
+            If both 'id' and 'name' are provided.
+        ValueError
+            If no auto moderation rule is found with the given name.
```
Comment (Lulalaby) [inline discord/guild.py:3846]:
```suggestion
```

File: discord/guild.py [hunk 4]
```diff
@@ -3816,3 +3816,58 @@ async def create_auto_moderation_rule(
...
+            # Get all auto moderation rules
+            rules_response = await self._state.http.get_auto_moderation_rules(self.id)
+            rules = rules_response
+
+            # Find the rule by name
+            matching_rules = [rule for rule in rules if rule["name"] == name]
+            if not matching_rules:
+                raise ValueError(f"No auto moderation rule found with name '{name}'.")
+            id = matching_rules[0]["id"]
+
```
Comment (Lulalaby) [inline discord/guild.py:3869]:
```suggestion
```

File: discord/guild.py [hunk 5]
```diff
@@ -3816,3 +3816,58 @@ async def create_auto_moderation_rule(
             self.id, payload, reason=reason
         )
         return AutoModRule(state=self._state, data=data)
+
+    async def delete_auto_moderation_rule(
+        self,
+        *,
+        id: int | None = None,
```
Comment (Lulalaby) [inline discord/guild.py:3823]:
```suggestion
        id: int,
```

**Review by Lulalaby — CHANGES_REQUESTED** [review]

File: discord/guild.py [hunk 6]
```diff
@@ -3816,3 +3816,32 @@ async def create_auto_moderation_rule(
...
+        HTTPException
+            Deleting the auto moderation rule failed.
+        Forbidden
+            You do not have the Manage Guild permission.
+        """
+
+        data = await self._state.http.delete_auto_moderation_rule(
+            self.id, id, reason=reason
+        )
+        return
```
Comment (Lulalaby) [inline discord/guild.py:3847]:
```suggestion
        return AutoModRule(state=self._state, data=data)
```

File: discord/guild.py [hunk 7]
```diff
@@ -3816,3 +3816,32 @@ async def create_auto_moderation_rule(
...
+        *,
+        id: int,
+        reason: str | None = None,
+    ) -> AutoModRule:
+        """
+        Deletes an auto moderation rule.
+
+        Parameters
+        ----------
+        id: Optional[int]
```
Comment (Lulalaby) [inline discord/guild.py:3831]:
```suggestion
        id: int
```

File: discord/guild.py [hunk 8]
```diff
@@ -3816,3 +3816,32 @@ async def create_auto_moderation_rule(
...
+        *,
+        id: int,
+        reason: str | None = None,
+    ) -> AutoModRule:
+        """
+        Deletes an auto moderation rule.
+
+        Parameters
+        ----------
+        id: int
```
Comment (Lulalaby) [inline discord/guild.py:3831]:
```suggestion
        id: :class:`int`
```

File: discord/guild.py [hunk 9]
```diff
@@ -3816,3 +3816,32 @@ async def create_auto_moderation_rule(
...
+            The ID of the auto moderation rule.
+        reason: Optional[str]
```
Comment (Lulalaby) [inline discord/guild.py:3833]:
```suggestion
        reason: Optional[:class:`str`]
```

**Review by Lulalaby — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 10]
```diff
@@ -12,6 +12,8 @@ These changes are available on the `master` branch, but have not yet been releas
 
 ### Added
 
+- Added function `delete_auto_moderation_rule`.
+  ([#2153]https://github.com/Pycord-Development/pycord/pull/2153)
```
Comment (plun1331) [inline CHANGELOG.md:16]: Bad markdown
```suggestion
  ([#2153](https://github.com/Pycord-Development/pycord/pull/2153))
```

File: discord/guild.py [hunk 11]
```diff
@@ -3816,3 +3816,32 @@ async def create_auto_moderation_rule(
             self.id, payload, reason=reason
         )
         return AutoModRule(state=self._state, data=data)
+
+    async def delete_auto_moderation_rule(
+        self,
+        *,
+        id: int,
```
Comment (plun1331) [inline discord/guild.py:3823]: `id` should be positional
```suggestion
        id: int,
        *,
```

File: discord/guild.py [hunk 12]
```diff
@@ -3816,3 +3816,32 @@ async def create_auto_moderation_rule(
...
+        HTTPException
+            Deleting the auto moderation rule failed.
+        Forbidden
+            You do not have the Manage Guild permission.
+        """
+
+        data = await self._state.http.delete_auto_moderation_rule(
+            self.id, id, reason=reason
+        )
+        return AutoModRule(state=self._state, data=data)
```
Comment (plun1331) [inline discord/guild.py:3847]: Deleting a rule does not return anything (status code 204)
https://discord.com/developers/docs/resources/auto-moderation#delete-auto-moderation-rule
Update the return typehint as well

**Review by plun1331 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 13]
```diff
@@ -12,6 +12,8 @@ These changes are available on the `master` branch, but have not yet been releas
 
 ### Added
 
+- Added function `delete_auto_moderation_rule`.
```
Comment (JustaSqu1d) [inline CHANGELOG.md:15]:
```suggestion
- Added function `Guild.delete_auto_moderation_rule`.
```

Comment (JustaSqu1d) [reply CHANGELOG.md:15]: The entry should also be appended to the bottom of the "Added" section, not the top.

**Review by JustaSqu1d — CHANGES_REQUESTED** [review]

**Review by JustaSqu1d — COMMENTED** [review]
Maybe add `AutoModRule.delete()` as well?

**Review by Lulalaby — APPROVED** [review]

**Pull request merged**